lib/core_extensions.rb

```diff
@@ -39,9 +39,17 @@ def initialize(*attribute)
 
     def before_destroy(record)
       klasses.each do |klass|
-        record.send(klass.to_sym).each do |what|
-          what = what.to_label unless what.is_a? String
-          record.errors.add :base, _("%{record} is used by %{what}") % { :record => record, :what => what }
+        record.association(klass.to_sym).association_scope.each do |what|
+          error_message = _("%{record} is used by %{what}")
+          unless what.is_a? String
+            if what.class.find_by_id(what.id)
```

Comment:
Seems this is only going to work with taxonomies and not the usual authorisation system.  I had a user that could only view host groups, but is able to see the hostnames of hosts that it didn't have view_hosts permissions over.

It seems like this should call .authorized, which is actually very similar to the problem that #2564 is trying to solve elsewhere in the UI... not sure if that should be combined.